### PR TITLE
adds lexsort for arrays with object dtype

### PR DESCRIPTION
**Pull request by behzadnouri**
on master:

``` python
>>> a = np.array(['a', 'c', 'b'], dtype='O')
>>> np.argsort(a, kind='mergesort')
array([0, 2, 1])
>>> np.lexsort((a,))
TypeError: merge sort not available for item 0
```

though i did some tests, it needs to be verified that this will not leak when it makes copy.

### CODE REVIEW

**Comment by jaimefrio** [comment]
Nice catch! And the implementation LGTM: `_unaligned_strided_byte_copy` copies the data but leaves the refcounts alone, so I don't think there is any reference leak risk. Would be nice to have a test of lexsorting with some objects involved, though.

File: numpy/core/src/multiarray/item_selection.c [hunk 1]
```diff
@@ -1427,9 +1427,11 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
                 goto fail;
             }
         }
-        if (!PyArray_DESCR(mps[i])->f->argsort[NPY_MERGESORT]) {
+        if (!PyArray_DESCR(mps[i])->f->argsort[NPY_MERGESORT]
+                && !PyArray_DESCR(mps[i])->f->compare) {
```
Comment (charris) [inline numpy/core/src/multiarray/item_selection.c]: Should this be `||`?

Comment (jaimefrio) [inline numpy/core/src/multiarray/item_selection.c]: It could be `!(a || b)`, which of course is the same as the current `!a && !b`.

**Comment by behzadnouri** [comment]
added tests

**Comment by charris** [comment]
Needs a note in `doc/release/1.11.0-notes.rst` under `Improvements`. The commit message should begin `ENH: `, see `doc/source/dev/gitwash/development_workflow.rst` for commit message instruction. You can use `git rebase -i HEAD~1` to rewrite the commit message, followed by `git push --force origin <your-branch>`.

**Comment by behzadnouri** [comment]
@charris  done

**Comment by charris** [comment]
Thanks @behzadnouri.

**Pull request merged**